tags/TagMap.md

```diff
@@ -0,0 +1,125 @@
+32 and 126)
+
+## TagScope
+
+`TagScope` is used to determine the scope of a `Tag`. The values for the `TagScope` are
+Local or Request. In future, additional values can be added to address specific situations.
+
+The tag creator determines the scope of the tag.
+
+**Local Scope**
+Tag with `Local` scope are used within the process it created. Such tags are not propagated
+across process boundaries. Even if the process is reentrant the tag MUST be excluded from
+propagation when the call leaves the process.
+
+**Request Scope**
+If a tag is created with the `Request` scope then it is propagated across process boundaries subject 
+to outgoing and incoming (on remote side) filter criteria. See `TagPropagationFilter` in 
+[Tag Propagation](#Tag Propagation). Typically 'Request' tags represents a request, processing
+ of which may span multiple entities.
+
+# TagMap 
+`TagMap` is an abstract data type that represents collection of tags. 
+i.e., each key is associated with exactly one value.  `TagMap` is serializable, and it represents
+all of the information that could be propagated inside the process and across process boundaries.  
+`TagMap` is a recommended name but languages can have more language specific name.
+
+## Limits
+Combined size of all `Tag`s should not exceed 8192 bytes before encoding.
+The size restriction applies to the deserialized tags so that the set of decoded
+ `TagMap`s is independent of the encoding format.
+
+## TagMap Propagation
+`TagMap` may be propagated across process boundaries or across any arbitrary boundaries for various 
+reasons. For example, one may propagate 'project-id' Tag across all micro-services to break down metrics
+by 'project-id'. Not all `Tag`s in a `TagMap` should be propagated and not all `Tag`s in a `TagMap`
+should be accepted from a remote peer. Hence, `TagMap` propagator must allow specifying an optional
+list of ordered `TagPropagationFilter`s for receiving `Tag`s or for forwarding `Tag`s or for both. 
+A `TagPropagationFilter` list for receiving MAY be different then that for forwarding.
+
+If no filter is specified for receiving then all `Tag`s are received. 
+If no filter is specified for forwarding then all `Tag`s are forwarded except those that have `Local Scope`.
+
+### TagPropagationFilter
+Tag Propagation Filter consists of action (`TagPropagationFilterAction`) and condition 
+(`TagPropagationFilterMatchOperator` and `TagPropagationFilterMatchString`). A `TagKey` 
+is evaluated against condition of each `TagPropagationFilter` in order. If the condition is evaluated
+to true then action is taken according to `TagPropagationFilterAction` and filter processing is stopped.
+If the condition is evaluated to false then the `TagKey` is processed against next `TagPropagationFilter`
+in the ordered list. If none of the condition is evaluated to true then the default
+action is **Exclude**.
+
+#### TagPropagationFilterAction
+There are two actions to apply to a Tag.
+
+**Exclude**
```

Comment:
I would make this an interface that way we can add for example sanitization or mutation of the TagValue. Include exclude is fine to be the first supported actions.